File: app/src/main/java/br/com/medeve/Activitys/CadastrarUsuarioActView.java
package br.com.medeve.Activitys;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;

import br.com.medeve.Controlers.UsuarioControler;
import br.com.medeve.Models.Usuario;
import br.com.medeve.R;
import br.com.medeve.Helpers.AccessFirebase;
import br.com.medeve.Models.Cliente;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Toast;


public class CadastrarUsuarioActView extends AppCompatActivity implements View.OnClickListener, RadioGroup.OnCheckedChangeListener {

    EditText edtNomeUsuario;
    EditText edtEmailUsuario;
    EditText edtSenhaUsuario;
    EditText edtConfirmaSenhaUsuario;

    Button btnCadastrarUsuario;

    RadioGroup rd_sexo;
    RadioButton rd_feminino, rd_masculino;

    String sexo;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cadastro_user);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        edtNomeUsuario = findViewById(R.id.ed_cad_user_nome);
        edtEmailUsuario = findViewById(R.id.ed_cad_user_email);
        edtSenhaUsuario = findViewById(R.id.ed_cad_user_senha);
        edtConfirmaSenhaUsuario = findViewById(R.id.ed_cad_user_confirmasenha);
        rd_sexo = findViewById(R.id.rd_sexos);
        btnCadastrarUsuario = findViewById(R.id.btn_cadastrar);
        rd_feminino = findViewById(R.id.femi);
        rd_masculino = findViewById(R.id.masc);

        setTitle("Cadastre-se");

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeButtonEnabled(true);

        btnCadastrarUsuario.setOnClickListener(this);
        rd_sexo.setOnCheckedChangeListener(this);
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onCheckedChanged(RadioGroup group, int checkedId) {
        switch (group.getId()) {
            case R.id.rd_sexos:
                if (rd_feminino.isChecked()) {
                    sexo = "Feminino";
                } else if (rd_masculino.isChecked()) {
                    sexo = "Masculino";
                }
                break;
        }
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_cadastrar:
                Usuario usuario = new Usuario();
                usuario.setSenha(edtSenhaUsuario.getText().toString());
                usuario.setConfirmaSenha(edtConfirmaSenhaUsuario.getText().toString());
                usuario.setEmail(edtEmailUsuario.getText().toString());
                usuario.setNome(edtNomeUsuario.getText().toString());
                usuario.setSexo(sexo);

                validarCadastroUsuario(usuario);
                break;
        }
    }


    private void validarCadastroUsuario(Usuario usuario) {
        if (usuario.nomeVazio()) {
            Toast.makeText(this, "Digite o seu nome", Toast.LENGTH_SHORT).show();
        } else if (usuario.senhaVazia()) {
            Toast.makeText(this, "Digite sua senha", Toast.LENGTH_SHORT).show();
        } else if (usuario.confirmarSenhaVazia()) {
            Toast.makeText(this, "Confirme sua senha", Toast.LENGTH_SHORT).show();
        } else if (usuario.nomeVazio()) {
            Toast.makeText(this, "Digite seu nome", Toast.LENGTH_SHORT).show();
        } else if (!usuario.getSenha().equals(usuario.getConfirmaSenha())) {
            Toast.makeText(this, "As senhas estão diferentes", Toast.LENGTH_SHORT).show();
        } else if (usuario.getSexo() == null){
            Toast.makeText(this, "Informe seu sexo", Toast.LENGTH_SHORT).show();
        } else {
            UsuarioControler.getInstance().cadastrar(usuario, CadastrarUsuarioActView.this);
        }
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                Intent i_cad_user = new Intent(getApplicationContext(), EntrarUsuarioActView.class);
                startActivity(i_cad_user);
                finish();
                break;
            default:
                break;
        }
        return true;
    }
}

File: app/src/main/java/br/com/medeve/Activitys/CadastrarUsuarioActView.kt
package br.com.medeve.Activitys

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import br.com.medeve.Helpers.ProgressBarHelper
import br.com.medeve.Models.Usuario
import br.com.medeve.R
import br.com.medeve.Util.Resultados
import br.com.medeve.ViewModels.UsuarioViewModel
import kotlinx.android.synthetic.main.activity_cadastro_user.*
import kotlinx.android.synthetic.main.activity_entrar_usuario.*
import org.koin.androidx.viewmodel.ext.android.viewModel

class CadastrarUsuarioActView : AppCompatActivity(), View.OnClickListener, RadioGroup.OnCheckedChangeListener {

    var sexo : String? = null

    val viewModelCadastroUsuario : UsuarioViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cadastro_user)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        title = "Cadastre-se"
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setHomeButtonEnabled(true)

        btn_cadastrar!!.setOnClickListener(this)
        rd_sexos!!.setOnCheckedChangeListener(this)

        setObservers()
    }

    override fun onCheckedChanged(group: RadioGroup, checkedId: Int) {
        when (group.id) {
            R.id.rd_sexos -> if (femi!!.isChecked) {
                sexo = "Feminino"
            } else if (masc!!.isChecked) {
                sexo = "Masculino"
            }
        }
    }

    @SuppressLint("NonConstantResourceId")
    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_cadastrar -> {
                val usuario = Usuario()
                usuario.senha = ed_cad_user_senha!!.text.toString()
                usuario.confirmaSenha = ed_cad_user_confirmasenha!!.text.toString()
                usuario.email = ed_cad_user_email!!.text.toString()
                usuario.nome = ed_cad_user_nome!!.text.toString()
                usuario.sexo = sexo
                validarCadastroUsuario(usuario)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                val i_cad_user = Intent(applicationContext, EntrarUsuarioActView::class.java)
                startActivity(i_cad_user)
                finish()
            }
            else -> {
            }
        }
        return true
    }

    private fun validarCadastroUsuario(usuario: Usuario) {
        if (usuario.nomeVazio()) {
            Toast.makeText(this, "Digite o seu nome", Toast.LENGTH_SHORT).show()
        } else if (usuario.senhaVazia()) {
            Toast.makeText(this, "Digite sua senha", Toast.LENGTH_SHORT).show()
        } else if (usuario.confirmarSenhaVazia()) {
            Toast.makeText(this, "Confirme sua senha", Toast.LENGTH_SHORT).show()
        } else if (usuario.nomeVazio()) {
            Toast.makeText(this, "Digite seu nome", Toast.LENGTH_SHORT).show()
        } else if (usuario.senha != usuario.confirmaSenha) {
            Toast.makeText(this, "As senhas estão diferentes", Toast.LENGTH_SHORT).show()
        } else if (usuario.sexo == null) {
            Toast.makeText(this, "Informe seu sexo", Toast.LENGTH_SHORT).show()
        } else {
            viewModelCadastroUsuario.cadastrarUsuario(usuario)
        }
    }

    private fun setObservers(){

        viewModelCadastroUsuario.resultadoCadastroUsuario.observe(this, {

            val progressBarHelper = ProgressBarHelper(this)
            when (it) {
                Resultados.CadastroUsuario.SUCESSO_CADASTRO_USUARIO -> {
                    Toast.makeText(this, "Usuário cadastrado com sucesso.", Toast.LENGTH_LONG).show()
                    //IntentHelper.instance!!.intentWithFinish(this,CadastroClienteActView::class.java)
                }
                Resultados.CadastroUsuario.EMAIL_JA_CADASTRADO -> {
                    Toast.makeText(this, "Email já cadastrado", Toast.LENGTH_LONG).show()
                    progressBarHelper.dismiss()
                }
                Resultados.CadastroUsuario.EMAIL_INVALIDO -> {
                    Toast.makeText(this, "Email inválido", Toast.LENGTH_LONG).show()
                    progressBarHelper.dismiss()
                }
                Resultados.CadastroUsuario.SENHA_COM_MENOS_DE_SEIS_CARACTERES -> {
                    Toast.makeText(this, "Senha inferior a 6 caracteres", Toast.LENGTH_LONG).show()
                    progressBarHelper.dismiss()
                }
                Resultados.CadastroUsuario.ERRO_DESCONHECIDO -> {
                    Toast.makeText(this, "Erro Desconhecido. Falha. Ao cadastar usuário", Toast.LENGTH_LONG).show()
                    progressBarHelper.dismiss()
                }
            }
        })
    }
}